addons/mail/controllers/google_translate.py

```diff
@@ -0,0 +1,58 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import requests
+import babel
+from odoo.http import request, route, Controller
+
+
+class GoogleTranslateController(Controller):
+    @route("/mail/message/translate", type="json", auth="user")
+    def translate(self, message_id):
+        user_lang = request.env.user.lang.split("_")[0]
+        translation = (
+            request.env["mail.message.translation"]
+            .sudo()
+            .search(
+                [("message_id", "=", message_id), ("target_lang", "=", user_lang)],
+                limit=1,
+            )
+        )
+
+        if not translation:
+            try:
+                body = str(request.env["mail.message"].browse(message_id).body)
+                code = self._detect_source(body)
+                source = babel.Locale(code).get_display_name(request.env.user.lang)
+                value = self._get_translation(body, code, user_lang)
+                translation = (
+                    request.env["mail.message.translation"]
+                    .sudo()
+                    .create(
+                        {
+                            "message_id": message_id,
+                            "target_lang": user_lang,
+                            "body": value,
+                            "source_lang": source,
+                        }
+                    )
+                )
+
+            except requests.exceptions.HTTPError as err:
+                return {"error": err.response.json()["error"]["message"]}
+
+        return {"source": translation.source_lang, "value": translation.body}
```

Comment:
I would suggest keeping the same key name everywhere for source_lang, as well as for body (value)